File: core/src/androidTest/java/com/emarsys/core/feature/FeatureRegistryTest.java
package com.emarsys.core.feature;

import com.emarsys.core.api.experimental.FlipperFeature;
import com.emarsys.testUtil.TimeoutUtils;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TestRule;

import java.util.Arrays;
import java.util.List;

import static junit.framework.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class FeatureRegistryTest {

    private FlipperFeature feature1;
    private FlipperFeature feature2;
    private FlipperFeature feature3;
    private List<FlipperFeature> features;

    @Rule
    public TestRule timeout = TimeoutUtils.getTimeoutRule();

    @Before
    public void setUp() {
        FeatureRegistry.reset();

        feature1 = mock(FlipperFeature.class);
        when(feature1.getFeatureName()).thenReturn("feature1");

        feature2 = mock(FlipperFeature.class);
        when(feature2.getFeatureName()).thenReturn("feature2");

        feature3 = mock(FlipperFeature.class);
        when(feature3.getFeatureName()).thenReturn("feature3");

        features = Arrays.asList(feature1, feature2, feature3);
    }

    @After
    public void tearDown() {
        FeatureRegistry.reset();
    }

    @Test
    public void testIsFeatureEnabled_shouldDefaultToBeingTurnedOff() {
        for (FlipperFeature feature : features) {
            assertFalse(FeatureRegistry.isFeatureEnabled(feature));
        }
    }

    @Test
    public void testIsFeatureEnabled_shouldReturnTrue_whenFeatureIsTurnedOn() {
        FeatureRegistry.enableFeature(feature1);
        assertTrue(FeatureRegistry.isFeatureEnabled(feature1));
    }

    @Test
    public void testEnableFeature_shouldAppendFeaturesToTheEnabledFeatureSet() {
        assertEquals(0, FeatureRegistry.enabledFeatures.size());
        FeatureRegistry.enableFeature(feature1);
        FeatureRegistry.enableFeature(feature2);
        assertEquals(2, FeatureRegistry.enabledFeatures.size());
    }

    @Test
    public void testDisableFeature_shouldRemoveFeaturesFromTheEnabledFeatureSet() {
        FeatureRegistry.enableFeature(feature1);
        FeatureRegistry.enableFeature(feature2);
        assertEquals(2, FeatureRegistry.enabledFeatures.size());
        FeatureRegistry.disableFeature(feature1);
        assertEquals(1, FeatureRegistry.enabledFeatures.size());
    }

    @Test
    public void testReset_shouldRemoveAllFeaturesFromTheEnabledFeatureSet() {
        FeatureRegistry.enableFeature(feature1);
        FeatureRegistry.enableFeature(feature2);
        assertEquals(2, FeatureRegistry.enabledFeatures.size());
        FeatureRegistry.reset();
        assertEquals(0, FeatureRegistry.enabledFeatures.size());
    }

}

File: core/src/androidTest/java/com/emarsys/core/feature/FeatureRegistryTest.kt
package com.emarsys.core.feature

import com.emarsys.core.api.experimental.FlipperFeature
import com.emarsys.core.feature.FeatureRegistry.disableFeature
import com.emarsys.core.feature.FeatureRegistry.enableFeature
import com.emarsys.core.feature.FeatureRegistry.isFeatureEnabled
import com.emarsys.core.feature.FeatureRegistry.reset
import io.kotest.matchers.shouldBe
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever

class FeatureRegistryTest  {
    private lateinit var feature1: FlipperFeature
    private lateinit var feature2: FlipperFeature
    private lateinit var feature3: FlipperFeature
    private lateinit var features: List<FlipperFeature>


    @Before
    fun setUp() {
        reset()
        feature1 = mock()
        whenever(feature1.featureName).thenReturn("feature1")
        feature2 = mock()
        whenever(feature2.featureName).thenReturn("feature2")
        feature3 = mock()
        whenever(feature3.featureName).thenReturn("feature3")
        features = listOf(feature1, feature2, feature3)
    }

    @After
    fun tearDown() {
        reset()
    }

    @Test
    fun testIsFeatureEnabled_shouldDefaultToBeingTurnedOff() {
        for (feature in features) {
            isFeatureEnabled(feature) shouldBe false
        }
    }

    @Test
    fun testIsFeatureEnabled_shouldReturnTrue_whenFeatureIsTurnedOn() {
        enableFeature(feature1)
        isFeatureEnabled(feature1) shouldBe true
    }

    @Test
    fun testEnableFeature_shouldAppendFeaturesToTheEnabledFeatureSet() {
        FeatureRegistry.enabledFeatures.size shouldBe 0
        enableFeature(feature1)
        enableFeature(feature2)
        FeatureRegistry.enabledFeatures.size shouldBe 2
    }

    @Test
    fun testDisableFeature_shouldRemoveFeaturesFromTheEnabledFeatureSet() {
        enableFeature(feature1)
        enableFeature(feature2)
        FeatureRegistry.enabledFeatures.size shouldBe 2
        disableFeature(feature1)
        FeatureRegistry.enabledFeatures.size shouldBe 1
    }

    @Test
    fun testReset_shouldRemoveAllFeaturesFromTheEnabledFeatureSet() {
        enableFeature(feature1)
        enableFeature(feature2)
        FeatureRegistry.enabledFeatures.size shouldBe 2
        reset()
        FeatureRegistry.enabledFeatures.size shouldBe 0
    }
}